(Partially) Closes #351: Javadocs for map APIs

jdimeo opened the pull request:
I have completed Javadocs on the API's `map` package. I am opening a PR to being the review process now before I (or someone else!) moves on to the other packages (`list`, `set`, `bag`, etc.).  In my view, `map` was the highest priority with methods like `getIfAbsentPutWith` that benefited most from Javadocs. Additionally, I'm envisioning many Javadocs in those packages will be copy-paste from what I have here, so I want to confirm the verbiage is right before doing so.

Some non-Javadoc changes included in this PR (which can be taken out if needed):
- Update checkstyle DTD (now content assist works in Eclipse when editing the configs)
- Remove `\r` checkstyle check. I don't see what benefit it provides (from what I can tell, my commits still have just `\n` even from Windows) while just making it a lot harder to contribute from Windows and get a clean build
- Remove a redundant `tap()` declaration in `immutableObjectPrimitiveMap.stg`

Additionally, this includes a helper `JavadocUtil` that will copy Javadocs back _from_ a generated class _to_ a template. It's not perfect but it saved me a ton of time, so that I could edit the Javadocs on an example concrete class and benefit from Eclipse's Javadoc assists that would otherwise be unavailable when directly editing the template.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/immutableObjectPrimitiveMap.stg:
```diff
@@ -39,22 +38,33 @@ public interface ImmutableObject<name>Map\<K> extends Object<name>Map\<K>
-    /**
-     * @since 9.0.
-     */
-    default ImmutableObject<name>Map\<K> tap(<name>Procedure procedure)
```
Hi @jdimeo, why is this method being removed?

jdimeo replied:
@donraab it's redundant with/trivially overrides `tap()` that is also declared in the parent interface `objectPrimitiveMap`.

donraab replied:
Is the type that is returned here different then in the parent interface? If so, this is a breaking change. This should also not be grouped together with the great JavaDoc work you are doing here as it is unrelated. 

Please revert it and if it is truly a duplicate we can remove it in a separate PR.

jdimeo replied:
@donraab sounds good I will revert (pending other feedback). Even if the return type is changing, I think we should implement it by calling the super default impl with an `@Override` instead of reimplementing

donraab replied:
Thanks @jdimeo. We've had some discussion on how best to address issues like this. While I agree that calling super is conceptually the better thing to do, in practice for a single method call implementation like forEach, it winds up being less verbose duplicating the method call. This is non-obvious at first. I had to try it to see myself. If there is more than a single method call then calling super is definitely the only reasonable option. It bothers me to have even a single line of duplicate code, but this is a case where I've learned to deal with it.

jdimeo replied:
Thanks @donraab - I see it differently, because if down the road if I change the default implementation in the super interface, I would expect all subclasses/sub-interfaces to use my new implementation (unless they meaningfully overrode it) and not want to have to track down every duplicated re-implementation. Verbosity (or lack thereof) isn't at all what I was going for. But of _course_ I'm happy to defer to you all as maintainers of this code base!

donraab replied:
We see things quite similarly I think. I've just allowed myself some flexibility in the cases of single method default methods where the likelihood of there ever being a change is as close to zero as possible, and the cost in verbosity or worse, casting is non zero. I sometimes save my argument points. I'll see if I can dig up the discussion thread we recently had on this and I'll link it here and extract the approach into some docs.

donraab replied:
@jdimeo @motlin @nikhilnanivadekar Here's the recent discussion on this topic of one line duplication vs. super calls for reference here: https://github.com/eclipse/eclipse-collections/pull/862#pullrequestreview-392463207

donraab replied:
@jdimeo I will try and read through all of the Javadoc additions in your PR by the weekend. Thank you for this terrific contribution, it is much appreciated.

jdimeo replied:
Thanks @donraab - no rush as it took me over 3 years to do this much ;-) you and others have been very responsive. thank you!

donraab commented on checkstyle-configuration.xml:
```diff
@@ -334,9 +334,4 @@
         <property name="message" value="Extra newline." />
     </module>
-
-    <module name="RegexpMultiline">
```
@motlin @nikhilnanivadekar Are you both ok with this removal?

motlin replied:
Is this for windows development? I'm confused because even when I used to develop on windows, this rule didn't fire for me. I don't feel like digging into why, so I'm ok with removing it.

jdimeo replied:
@motlin I think it's because a) I'm on Windows and b) my Git is configured to check out files with platform-specific line endings. However, c) my Git is _also_ configured to check in Linux `\n` line endings, hence my commits pass on the server/didn't change anything. I feel like explicitly blacklisting `\r` makes development more difficult for new contributors and/or Windows users when they have the experience that I had where the build failed after cloning and doing nothing

donraab replied:
@motlin @jdimeo Someone else was complaining on collections-dev mail group they were having issues with checkstyle on Windows. Maybe related. I'm ok with this change.

nikhilnanivadekar replied:
I tried to fix this checkstyle configuration in my local. After numerous attempts I am at a dead-end to have something work predictable. I am fine removing this change. Thanks @jdimeo 
Tagging @motlin and @donraab for transparency.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitiveObjectMap.stg:
```diff
@@ -45,20 +53,71 @@ public interface Mutable<name>ObjectMap\<V> extends <name>ObjectMap\<V>, Mutable
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates a value with the key.
+     * a new value with they key
```
This sentence fragment looks like it might be superfluous.

jdimeo replied:
Good catch! I'll fix it.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitiveObjectMap.stg:
```diff
@@ -45,20 +53,71 @@ public interface Mutable<name>ObjectMap\<V> extends <name>ObjectMap\<V>, Mutable
 
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates the result of invoking the value supplier with the key.
```
I don't believe the key is passed to the function, as it is a `Function0` which means it takes no paramters.

jdimeo replied:
That is what I meant, but I agree it's a bit confusing. By saying "value supplier" I was hoping that readers would realize a `Function0` == `Supplier<>` from Java 8.  I was trying to say "associates ... with the key" not "associates ... invoking ... with the key". I'll try to word it more clearly.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitiveObjectMap.stg:
```diff
@@ -45,20 +53,71 @@ public interface Mutable<name>ObjectMap\<V> extends <name>ObjectMap\<V>, Mutable
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates the result of invoking the value function with the key.
+     * @param key the key
```
The value function is invoked with the additional parameter here, not the key.

jdimeo replied:
P.S. I feel like those parameters/methods overloads are largely unnecessary because a function with a fixed parameter is basically the same thing as a consumer.

nikhilnanivadekar replied:
@jdimeo while that is true, this is a part of micro-optimization. Basically, when you pass in a parameter, you can cache it for that loop and avoid things like finality and reassignment issues. Secondly, the `*with` patterns help us use Method References.

jdimeo replied:
Understood. In general, I'm a big fan of how your APIs are not super spartan and provide some "unnecessary" conveniences like these, so 👍

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitivePrimitiveMap.stg:
```diff
@@ -56,20 +68,88 @@ public interface Mutable<name1><name2>Map extends <name1><name2>Map, Mutable<nam
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates a value with the key.
+     * a new value with they key
```
A sentence fragment.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitivePrimitiveMap.stg:
```diff
@@ -56,20 +68,88 @@ public interface Mutable<name1><name2>Map extends <name1><name2>Map, Mutable<nam
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates the result of invoking the value supplier with the key.
+     * @param key the key
```
I understand this better after reading your comment from before but could still be confusing.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitivePrimitiveMap.stg:
```diff
@@ -56,20 +68,88 @@ public interface Mutable<name1><name2>Map extends <name1><name2>Map, Mutable<nam
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates the result of invoking the value function with the key.
+     * @param key the key
```
Interestingly, with "with the key" has dual meaning as the value is associated with the key, and they key is passed as a parameter to the function. I think this is ok here.

donraab commented on eclipse-collections-code-generator/src/main/resources/api/map/mutablePrimitivePrimitiveMap.stg:
```diff
@@ -56,20 +68,88 @@ public interface Mutable<name1><name2>Map extends <name1><name2>Map, Mutable<nam
+    /**
+     * Retrieves the value associated with the key if one exists; if it does not,
+     * associates the result of invoking the value function with the key.
+     * @param key the key
```
This is where it gets confusing. The calculated value is associated with the key, but the parameter is passed to the function.

jdimeo commented:
@donraab thank you for the review, feedback, and encouragement! I saw this earlier in the week and wanted my next comment to be another push with the requested changes, but I was unable to get to this this week. It's on the to-do list for next week, and I'll see if I can knock out a couple other packages as well (`set`, `bag`, etc.)

nikhilnanivadekar commented:
I would suggest let’s get this PR merged before you add more packages. That way the changes and commits are small and reviews and feedback can be faster

nikhilnanivadekar commented:
@jdimeo I would like to close out the feature set for 10.3.0 and release it soon. That way we can open the major version 11.0.0 target. If possible, let us try to get this PR in as Javadocs help us throughout the development experience. Once all the comments suggested by @donraab  are addressed, I will need your (@jdimeo ) help in a bit of house keeping especially to squash the merge commits and commits which have just stylistic changes. We can visit them once all the changes are approved. We are super close, and it would be great to have this merged. Thanks a ton for this valuable contribution!

jdimeo commented:
@donraab @nikhilnanivadekar I've pushed a commit that addresses the review items.  Don't you all squash when you merge the PR? I've reverted the `tap()` change so the only non-Javadoc thing remaining are the few minor Checkstyle things. Was that what you meant as far as needing me to do housekeeping?

Happy to contribute to a wiki or blog. Just let me know where/how to submit a draft for review (I see that it appears I can directly edit the wiki, but I assume you would want to review it first?)

nikhilnanivadekar commented:
@jdimeo we normally like to squash commits, however, the changes in terms of LOC are pretty high, and I would like you to get the proper number of commit credit where it is due. So, let me elaborate the housekeeping things which I think we need to do before having a final pass review:

1. Fix-up https://github.com/eclipse/eclipse-collections/pull/887/commits/9d46f1174ebf3bff68d9be21f4bebf37a099f3e5 on https://github.com/eclipse/eclipse-collections/pull/887/commits/712cc712704f20d35164ae435d8298b128475a36

2. Rearrange and fix-up  https://github.com/eclipse/eclipse-collections/pull/887/commits/1719ab22d2e65f49cc7ec7a1bd0c15047491f3d6 on https://github.com/eclipse/eclipse-collections/pull/887/commits/ee6ea03af7fa9c61717f1e5cd3e8c5e0f90c57d0

3. Fix-up https://github.com/eclipse/eclipse-collections/pull/887/commits/1b8fcec6976edcdf51807368863239e0c2fbaf47 on https://github.com/eclipse/eclipse-collections/pull/887/commits/e0bdfc498a534897952db5994e960f2222570a53

4. Fix-up https://github.com/eclipse/eclipse-collections/pull/887/commits/c3682f014eb4555b750fa3c44c812ebc4e279a99 on https://github.com/eclipse/eclipse-collections/pull/887/commits/e0bdfc498a534897952db5994e960f2222570a53

5. Fix-up https://github.com/eclipse/eclipse-collections/pull/887/commits/8637734be775746470f07ad83988ea442cb24007 on https://github.com/eclipse/eclipse-collections/pull/887/commits/e0bdfc498a534897952db5994e960f2222570a53

6. Fix-up https://github.com/eclipse/eclipse-collections/pull/887/commits/23d1ffe3245ee7fd30619f636b6f54e673095d42 on https://github.com/eclipse/eclipse-collections/pull/887/commits/b08407afc3d18a0f71eaf517e91d8c087203c712

7. Drop this commit: https://github.com/eclipse/eclipse-collections/pull/887/commits/9cea83f2f420cc513edfea70b6084030e977b51d
I have a fix for this in my local and I will test and push it.

8. Fix up  https://github.com/eclipse/eclipse-collections/pull/887/commits/431b0be50bb9641d6a4bd6497ebdcca33889356c on https://github.com/eclipse/eclipse-collections/pull/887/commits/712d5cdb066375a63a631d23936fd5cb42a8b760 I don't want to be pedantic and make you split your change and fix-up the review comment changes on that particular commit hash. Instead I am fine if you fix-up the review comment changes commit to the last commit.

9. I did not see the commit which reverts the changes made for `tap()` but I might have missed something. So, if the `tap()` change is not reverted please revert it and fix-up on that particular commit where `tap()` was changed, so that the net effect looks to be zero. If you have already done this, sorry for missing that part.

10. Rebase on to upstream/master

After this we will do a final review, merge your PR and then I will cut a milestone release for testing. 
Thanks again for your contribution!

jdimeo commented:
@nikhilnanivadekar I've actually never done a "fix up" commit before since I usually work on smaller projects.  But I understand now how that works with autosquash. I personally would've been fine with a massive "Javadoc all the things" commit, but this is definitely better. I'll address it tomorrow (I hope!)

P.S. I did revert the tap change so the net effect is zero (I snuck in the commit between donraab's first review comment and all the others so it's easy to miss) . Are you OK with me adding the `@Override` tag? The only reason why I noticed that method in the first place was because I got warnings in Eclipse about it overridding but not being tagged with the annotation.

nikhilnanivadekar commented:
@jdimeo found the hidden commit for revert. So, can you fix-up https://github.com/eclipse/eclipse-collections/pull/887/commits/e36ca664abb49ed5fc49f24d1744bc582e03fbdc on https://github.com/eclipse/eclipse-collections/pull/887/commits/391e4fc74e08294ca210a4d0f0598f81a3627fc5

nikhilnanivadekar commented:
@jdimeo If you have `squash`ed commits you have essentially done a `fix-up`. Fix up is nothing but squash where the commit message from previous commit is picked. For additional documentation take a look here:
1. https://git-scm.com/docs/git-rebase
2. https://stackoverflow.com/a/16758662
3. https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
4. https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase 

If you do an interactive rebase (points 3, 4 above), git gives you the chance to move around commits. Rebase can be dangerous, especially when doing it for the first time as it is essentially rewriting history. So I would suggest to take a back up of the files before doing the rebase. 
Feel free to reach out with questions or concerns and I will be happy to help you through.

nikhilnanivadekar commented:
You can add the Override tag, but please add it as a separate commit for tracking. It just helps me while I write Release Notes.

jdimeo commented:
@nikhilnanivadekar doing the fixup of https://github.com/eclipse/eclipse-collections/commit/23d1ffe3245ee7fd30619f636b6f54e673095d42 created a weird merge conflict in the middle of the interactive rebase that I wasn't sure how to correctly resolve, Do I need to figure that out or can we roll with every other change except that one?

nikhilnanivadekar commented:
Let’s roll with the others and then revisit this. I can help you offline with that one later

jdimeo commented:
Thanks @nikhilnanivadekar I just forced push my history rewrite. Please tell me if I did it correctly! I also rebased on `upstream/master` so it _should_ be ready to go

nikhilnanivadekar commented:
@jdimeo actually now try again doing the fix up of https://github.com/eclipse/eclipse-collections/pull/887/commits/0798268fc859515ebb90e89db7dc997fc58aaf79 on https://github.com/eclipse/eclipse-collections/pull/887/commits/1e780c41a1d9750949b5a1a59470d4d38757643c
That should most likely go through. Give it a go once, please?

nikhilnanivadekar commented:
Also a minor nit pick, can you reword this commit: https://github.com/eclipse/eclipse-collections/pull/887/commits/a9f5d4bd9f13a91aec1336e73b020b7475e7bee7
to call it Javadocs to make it consistent with other commits. We are big fans of symmetry as you would see 😄  thanks again for your patience and cooperation.

jdimeo commented:
@nikhilnanivadekar I allocated a `short` to track your patience quota with me but it just overflowed to an `int`- I don't know how you do it! thank you! :-P

jdimeo commented:
@nikhilnanivadekar Yes! it worked this time no merge required. I also caught a missing " for " and added some `.` so they are all completely symmetrical now :-)

nikhilnanivadekar commented:
@jdimeo I missed a couple more things, sorry!
1. Reword https://github.com/eclipse/eclipse-collections/pull/887/commits/b0aa722ff72c68942cd38384391c059f2fcf7bc5 to keep only the change of Javadoc
2. Reword https://github.com/eclipse/eclipse-collections/pull/887/commits/20de3bf81417bf88e5b11d7fbdface7610bfa7d7 to make it succinct to convey only the removal for the check

jdimeo commented:
@nikhilnanivadekar done (I hope!)

jdimeo commented:
@nikhilnanivadekar Just confirming you don't need anything else from me. Also, I noticed this in `.gitattributes`:
```
# Auto detect text files and perform LF normalization
* text=auto
```
that could explain why we can safely remove the `\r` checkstyle check

nikhilnanivadekar commented on eclipse-collections-code-generator/src/main/resources/api/map/primitiveObjectMap.stg:
```diff
@@ -32,33 +32,92 @@ import org.eclipse.collections.api.tuple.primitive.<name>ObjectPair;
 
     /**
+     * Returns a view of the keys in this map. This iterable is backed by the map, so
+     * any modifications to the underlying map will be reflected in the keys returned
```
We might want to update the wording a bit in a future iteration. But it is not super important. 
`keysView()` `keyValueView()` essentially all `*View()` APIs are supposed to be lazy. So, the idea of changes in the map reflecting in the view might be a bit awkward. 
It is ok for now

nikhilnanivadekar commented on eclipse-collections-code-generator/src/test/java/org/eclipse/collections/codegenerator/tools/JavadocUtil.java:
```diff
@@ -0,0 +1,206 @@
+            new JavadocUtil().generatedClass(args[i]).template(args[i + 1]).process();
+        }
+    }
+}
```
This utility is awesome! May be if it is possible, can you contribute another utility which will move code to the template? It will be super useful to be able to write code in one of the primitive files and reverse engineer it to a template. It need not be fully polished and working, but should provide a starting point. Thoughts?

jdimeo replied:
@nikhilnanivadekar I haven't forgotten about this request, but regretfully I'm not sure when I'll have the time to tackle it!

nikhilnanivadekar commented:
@jdimeo sorry didnt get a chance during the week to work on this. Finally got some time and did a final review. Looks good to me! Thanks a lot for your contribution!
If you get a moment please feel free to blog about the JavadocUtil and/or contributing to the project. 
Looking forward to your next PR. Cheers!

jdimeo commented:
@nikhilnanivadekar No worries! just wanted to make sure I wasn't the blocker. Happy to help.
Where/how would I submit a blog post draft?

nikhilnanivadekar commented:
I meant a blog in your personal space to showcase your experience. May be on Wordpress or even GitHub pages or Medium?
We don’t have an Eclipse Collections blog space, yet. But now that I think about it we should have one. Please don’t wait for us to set that one up.
